storage: TestInitRaftGroupOnRequest failed

[(storage).TestInitRaftGroupOnRequest failed](https://viewLog.html?buildId=1700915&tab=buildLog) on [master@69e0b0d2a685beae726904beda0713e4c3b61862](https://github.com/cockroachdb/cockroach/commits/69e0b0d2a685beae726904beda0713e4c3b61862):

```
=== RUN   TestInitRaftGroupOnRequest
I200120 19:43:55.588389 412000 gossip/gossip.go:395  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:44963" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 cluster_name:"" sql_address:<network_field:"" address_field:"" > 
W200120 19:43:55.608313 412000 gossip/gossip.go:1518  [n2] no incoming or outgoing connections
I200120 19:43:55.608402 412000 gossip/gossip.go:395  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:35827" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 cluster_name:"" sql_address:<network_field:"" address_field:"" > 
I200120 19:43:55.609309 412375 gossip/client.go:124  [n2] started gossip client to 127.0.0.1:44963
I200120 19:43:55.657882 412000 storage/replica_command.go:406  [s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/50 [r2] (manual)
I200120 19:43:55.679670 412000 storage/replica_command.go:1706  [s1,r2/1:/{Table/50-Max}] change replicas (add [(n2,s2):2LEARNER] remove []): existing descriptor r2:/{Table/50-Max} [(n1,s1):1, next=2, gen=1]
I200120 19:43:55.699920 412000 storage/replica_raft.go:248  [s1,r2/1:/{Table/50-Max},txn=f064b623] proposing SIMPLE(l2) ADD_REPLICA[(n2,s2):2LEARNER]: after=[(n1,s1):1 (n2,s2):2LEARNER] next=3
I200120 19:43:55.703041 412000 storage/store_snapshot.go:973  [s1,r2/1:/{Table/50-Max}] sending LEARNER snapshot a3d2c97f at applied index 14
I200120 19:43:55.703178 412000 storage/store_snapshot.go:1016  [s1,r2/1:/{Table/50-Max}] streamed snapshot to (n2,s2):2LEARNER: kv pairs: 6, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I200120 19:43:55.705322 412723 storage/replica_raftstorage.go:785  [s2,r2/2:{-}] applying LEARNER snapshot [id=a3d2c97f index=14]
I200120 19:43:55.706978 412723 storage/replica_raftstorage.go:806  [s2,r2/2:/{Table/50-Max}] applied LEARNER snapshot [total=2ms ingestion=4@1ms id=a3d2c97f index=14]
I200120 19:43:55.708497 412000 storage/replica_command.go:1706  [s1,r2/1:/{Table/50-Max}] change replicas (add [(n2,s2):2] remove []): existing descriptor r2:/{Table/50-Max} [(n1,s1):1, (n2,s2):2LEARNER, next=3, gen=2]
I200120 19:43:55.735164 412000 storage/replica_raft.go:248  [s1,r2/1:/{Table/50-Max},txn=4c8c5660] proposing SIMPLE(v2) ADD_REPLICA[(n2,s2):2]: after=[(n1,s1):1 (n2,s2):2] next=3
I200120 19:43:55.744371 412000 util/stop/stopper.go:539  quiescing
W200120 19:43:55.744531 412608 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/50/RangeDescriptor: node unavailable; try another peer
I200120 19:43:55.750320 412000 util/stop/stopper.go:539  quiescing
E200120 19:43:55.762440 412000 storage/client_raft_test.go:4015  expected raft group to be uninitialized
I200120 19:43:55.775424 413005 util/stop/stopper.go:539  quiescing
I200120 19:43:55.775489 413004 util/stop/stopper.go:539  quiescing
W200120 19:43:55.778077 412637 storage/raft_transport.go:294  unable to accept Raft message from (n2,s2):2: no handler registered for (n1,s1):1
I200120 19:43:55.778125 413003 util/stop/stopper.go:539  quiescing
W200120 19:43:55.778406 412758 storage/raft_transport.go:466  no handler found for store 2 in response range_id:2 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"store 1 was not found" transaction_restart:NONE origin_node:0 detail:<store_not_found:<store_id:1 > > now:<> > > 
W200120 19:43:55.778522 412724 storage/raft_transport.go:637  while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
W200120 19:43:55.778584 412685 storage/raft_transport.go:637  while processing outgoing Raft queue to node 2: EOF:
I200120 19:43:55.779063 413003 util/stop/stopper.go:539  quiescing
I200120 19:43:55.788775 413003 util/stop/stopper.go:539  quiescing
--- FAIL: TestInitRaftGroupOnRequest (0.34s)
    client_raft_test.go:4024: expected NotLeaseHolderError; got <nil>

```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestInitRaftGroupOnRequest PKG=./pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

[roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestInitRaftGroupOnRequest.%2A&sort=title&restgroup=false&display=lastcommented+project)

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -3914,6 +3914,12 @@ func TestInitRaftGroupOnRequest(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	storeCfg := kvserver.TestStoreConfig(nil /* clock */)
 	storeCfg.TestingKnobs.DisableMergeQueue = true
+	// Disable async intent resolution. This can lead to flakiness in the test
+	// because it allows for the intents written by the split transaction to be
+	// resolved at any time, including after the nodes are restarted. The intent
+	// resolution on the RHS's local range descriptor can both wake up the RHS
+	// range's Raft group and result in the wrong replica acquiring the lease.
+	storeCfg.TestingKnobs.IntentResolverKnobs.DisableAsyncIntentResolution = true
 	mtc := &multiTestContext{
 		storeConfig: &storeCfg,
 		// TODO(andrei): This test was written before multiTestContexts started with
@@ -3945,6 +3951,8 @@ func TestInitRaftGroupOnRequest(t *testing.T) {
 	mtc.restart()
 
 	// Get replica from the store which isn't the leaseholder.
+	// NOTE: StoreID is 1-indexed and storeIdx is 0-indexed, so despite what
+	// this might look like, this is grabbing the replica without the lease.
 	storeIdx := int(lease.Replica.StoreID) % len(mtc.stores)
 	if repl = mtc.stores[storeIdx].LookupReplica(roachpb.RKey(splitKey)); repl == nil {
 		t.Fatal("replica should not be nil for RHS range")

